### ISSUE TITLE
Replace Task.Result with await Task

### ISSUE THREAD

**Issue by AmadeusW** [code-cracker/code-cracker#734]
**New analyzer**: Replace `Task.Result` with `await` and make method `async Task`

Recently we've found some deadlocks in our code. They were caused by a stray `.Result` in situations when there was an `async` method earlier in the call stack. [Using `.Result` is considered a bad practice](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx). 
I propose an analyzer that removes `.Result` and replaces it with an `await ` call

Before:

````csharp
private bool doSomething()
{
    return asyncMethod().Result;
}
````

After:

````csharp
private Task<bool> doSomething()
{
    return await asyncMethod();
}
````

Category: Reliability or Design
Severity: By default hidden. I would probably change it to Warning in my own code

I'm curious what's your opinion on
1. This refactoring will cause a build error, because method's signature changes
2. Are there any use cases where using `.Result` is OK?

**Comment by giggio** [comment]
If we do this, we would need to change all the callers to also be async. We can't cause a build problem after. This means cascade into several parts of the program. This is a *huge* task.
I actually like it.
Using `.Result` is ok in several situations, when you can control the program flow. A console app that is calling into an async only API will *have* to `.Result`, for example.
Also, it would be a good idea to rename the function to end with `Async`.

Because of the cascading problem, I think we need to discuss it.

**Comment by AmadeusW** [comment]
Yes, this can introduce a huge change to the solution. Some functions can *not* be made async - for example, `Main` in console applications. Also, async event handlers that return `void` could crash the process with unhandled exceptions.

The more I think about this concept, the more I like it, too. After all, making an entire program async is a very pragmatic and menial task.

**Comment by SriramSakthivel** [comment]
If the method is already async and still uses `task.Result`, we could add a warning and provide code fix which converts the `task.Result` to `await task`

**Comment by AmadeusW** [comment]
I think this would be fair. This way, we won't cascade the conversion to async and introduce massive build errors.
It won't catch deadlocks caused by `async` method calling non-`async` method with `.Result`, but I think it would be a good start.
I can take a shot at implementing it if everyone is ok with this.

**Comment by giggio** [comment]
@AmadeusW I am ok with it, as long as the method already has the modifier `async`. So we can't change an async method that returns a Task directly, ok?
Later we can evolve this to be more ambitious and solve the original problem.
Go for the implementation, then, @AmadeusW.

**Comment by AmadeusW** [comment]
@giggio @ElemarJR @carloscds pinging you because the guidelines say that I should notify you before creating a PR.
I think I've got it working. Check it out at https://github.com/code-cracker/code-cracker/compare/master...AmadeusW:0122
Screenshot of a non trivial case:
![cc0122](https://cloud.githubusercontent.com/assets/1673956/16567981/5c63f2aa-422d-11e6-9859-5a45f5b5cbb8.png)
